File: frontend/src/types/banlist.ts
// Ban list / legality types

export interface BanListEntry {
  card_id: number;
  name_en: string | null;
  name_pt: string | null;
  set_code: string | null;
  collector_number: string | null;
  format: string;
  status: string;
  effective_date: string | null;
  image_url: string | null;
  printings: number;
  owned: boolean;
  owned_quantity: number;
}

export interface BanlistStatus {
  last_synced_at: string | null;
  legalities_count: number;
  banned_count: number;
  restricted_count: number;
  history_count: number;
  formats: number;
}

export interface CardLegality {
  format: string;
  status: string;
  effective_date: string | null;
}

export interface LegalityHistoryEntry {
  card_id: number;
  name_en: string | null;
  name_pt: string | null;
  set_code: string | null;
  collector_number: string | null;
  format: string;
  old_status: string | null;
  new_status: string;
  changed_at: string;
  image_url: string | null;
}

export interface LegalityHistoryResponse {
  items: LegalityHistoryEntry[];
  total: number;
  limit: number;
  offset: number;
}

export interface CardBanHistoryEntry {
  id: number;
  format: string;
  old_status: string | null;
  new_status: string;
  changed_at: string;
  source: string;
}

export interface BanImpactSchema {
  format: string;
  old_status: string | null;
  new_status: string;
  changed_at: string;
  window_days: number;
  price_before: number | null;
  price_after: number | null;
  absolute_change: number | null;
  percent_change: number | null;
  data_available: boolean;
}

File: src/api/schemas/banlist.py
"""Pydantic schemas for the banlist router."""

from __future__ import annotations

from datetime import date, datetime

from pydantic import BaseModel


class BanListEntry(BaseModel):
    card_id: int
    name_en: str | None = None
    name_pt: str | None = None
    set_code: str | None = None
    collector_number: str | None = None
    format: str
    status: str
    effective_date: date | None = None
    image_url: str | None = None


class CardLegalitySchema(BaseModel):
    format: str
    status: str
    effective_date: date | None = None


class LegalityHistoryEntry(BaseModel):
    card_id: int
    name_en: str | None = None
    name_pt: str | None = None
    set_code: str | None = None
    collector_number: str | None = None
    format: str
    old_status: str | None = None
    new_status: str
    changed_at: datetime
    image_url: str | None = None


class LegalityHistoryResponse(BaseModel):
    items: list[LegalityHistoryEntry]
    total: int
    limit: int
    offset: int


class CardBanHistoryEntry(BaseModel):
    id: int
    format: str
    old_status: str | None = None
    new_status: str
    changed_at: datetime
    source: str


class BanImpactSchema(BaseModel):
    format: str
    old_status: str | None = None
    new_status: str
    changed_at: datetime
    window_days: int
    price_before: float | None = None
    price_after: float | None = None
    absolute_change: float | None = None
    percent_change: float | None = None
    data_available: bool


class SyncTriggerRequest(BaseModel):
    bulk: bool = True
    limit: int | None = None
